code/modules/cargo/randomstock.dm

```diff
@@ -0,0 +1,1880 @@
+		var/minweight = 99999999999//We will distribute items somewhat evenly among crates
+		//by selecting the least-filled one for each spawn
+
+		for (var/obj/structure/closet/crate/C in containers)
+			if (C.stored_weight() < minweight && C.stored_weight() <= C.storage_capacity)
+				minweight = C.stored_weight()
+				emptiest = C
+				cratespawn = 1
+
+	if (cratespawn)
+		return emptiest
+	else
+		//If cratespawn is zero, we either failed to find a crate to spawn in, or didnt select
+		//crate spawning with the random check.
+		var/turf/clearest
+		var/min_items = 999999999
+		for (var/B in tables)//As with crates, we attempt to distribute items evenly. Pick least-filled
+			var/TB = get_turf(B)
+			var/numitems = 0
+			for (var/obj/item in TB)
+				numitems++
+			if (numitems <= min_items)
+				clearest = TB
+				min_items = numitems
+
+
+		if (!clearest)//This should never happen
+			clearest = get_turf(pick(tables))
+
+		return clearest
+
+
+/datum/cargospawner/proc/get_spawntype()
+	var/list/spawntypes = list("1" = STOCK_RARE_PROB, "2" = STOCK_UNCOMMON_PROB, "3" = (100 - (STOCK_RARE_PROB + STOCK_UNCOMMON_PROB)))
+	var/stocktype = pickweight(spawntypes)
+	switch (stocktype)
+		if ("1")
+			return pickweight(random_stock_rare)
+		if ("2")
+			return pickweight(random_stock_uncommon)
+		if ("3")
+			return pickweight(random_stock_common)
+
+//Minor and moderate mobs are checked per crate
+#define INFEST_PROB_MINOR	6
+#define INFEST_PROB_MODERATE	3
+
+#define INFEST_PROB_SEVERE	3//Severe is once per round, not per crate
+
+/datum/cargospawner/proc/handle_infestation()
+	for (var/obj/O in containers)
+		if (prob(INFEST_PROB_MINOR))
+	//No admin message for the minor mobs, because they are friendly and harmless
+			var/ctype = pickweight(infest_mobs_minor)
+			new ctype(O)
+		else if	 (prob(INFEST_PROB_MODERATE))
+			var/ctype = pickweight(infest_mobs_moderate)
+			new ctype(O)
+			msg_admin_attack("Common cargo warehouse critter [ctype] spawned inside [O.name] coords (<a href='?_src_=holder;adminplayerobservecoodjump=1;X=[O.x];Y=[O.y];Z=[O.z]'>JMP</a>)")
+
+
+	//This is checked only once per round. ~3% chance to spawn a scary monster infesting the warehouse
+	if (prob(INFEST_PROB_SEVERE))
+		//Find a tile to spawn the thing
+		var/list/turfs = list()
+		var/turf/T
+		for (var/turf/t in warehouse)
+			T = t//Failsafe incase none are clear
+			if (turf_clear(T))
+				turfs |= t
+
+		if (turfs.len)
+			T = pick(turfs)
+
+		var/ctype = pickweight(infest_mobs_severe)
+		if (ctype != "cratey")
+			new ctype(T)
+			msg_admin_attack("Rare cargo warehouse critter [ctype] spawned coords (<a href='?_src_=holder;adminplayerobservecoodjump=1;X=[T.x];Y=[T.y];Z=[T.z]'>JMP</a>)")
+			return
+		else
+			var/obj/C = pick(containers)
+			var/mob/living/simple_animal/hostile/mimic/copy/cratey = new /mob/living/simple_animal/hostile/mimic/copy(C.loc, C, null)
+			//Cratey is kinda tough but slow, easy to run away from
+			cratey.name = "Cratey"
```

Comment:
Didn't we generally agree cratey is a relic from the old days of adminbus. How can one justify this, lorewise and rpwise...